trieste/acquisition/rule.py

```diff
@@ -1488,46 +1502,70 @@ def get_initialize_subspaces_mask(
 
     def filter_datasets(
         self, models: Mapping[Tag, ProbabilisticModelType], datasets: Mapping[Tag, Dataset]
-    ) -> Mapping[Tag, Dataset]:
-        # Update subspaces with the latest datasets.
-        assert self._subspaces is not None
-        for subspace in self._subspaces:
-            # Re-initialize or update the subspace, depending on the property.
-            if subspace.requires_initialization:
-                subspace.initialize(models, datasets)
+    ) -> types.State[BatchTrustRegionState[UpdatableTrustRegionType] | None, Mapping[Tag, Dataset]]:
+        def state_func(
+            state: BatchTrustRegionState[UpdatableTrustRegionType] | None,
+        ) -> Tuple[BatchTrustRegionState[UpdatableTrustRegionType] | None, Mapping[Tag, Dataset]]:
+            if state is not None:
+                assert self._tags == state.subspace_tags, (
+                    f"The tags of the state acquisition space "
+                    f"{state.subspace_tags} should be the same as the tags of the "
+                    f"BatchTrustRegion acquisition rule {self._tags}"
+                )
+                assert len(state.subspaces) == len(state.subspace_tags), (
+                    f"Inconsistent number of subspaces: {len(state.subspaces)} subspaces"
+                    f"and {len(state.subspace_tags)} tags"
+                )
+                subspaces = tuple(state.subspaces)
             else:
-                subspace.update(models, datasets)
-        self.maybe_initialize_subspaces(self._subspaces, models, datasets)
-
-        # Filter out points that are not in any of the subspaces. This is done by creating a mask
-        # for each local dataset that is True for points that are in any subspace.
-        used_masks = {
-            tag: tf.zeros(dataset.query_points.shape[:-1], dtype=tf.bool)
-            for tag, dataset in datasets.items()
-            if LocalizedTag.from_tag(tag).is_local
-        }
-
-        for subspace in self._subspaces:
-            in_region_masks = subspace.get_datasets_filter_mask(datasets)
-            if in_region_masks is not None:
-                for tag, in_region in in_region_masks.items():
-                    ltag = LocalizedTag.from_tag(tag)
-                    assert ltag.is_local, f"can only filter local tags, got {tag}"
-                    used_masks[tag] = tf.logical_or(used_masks[tag], in_region)
-
-        filtered_datasets = {}
-        for tag, used_mask in used_masks.items():
-            filtered_datasets[tag] = Dataset(
-                tf.boolean_mask(datasets[tag].query_points, used_mask),
-                tf.boolean_mask(datasets[tag].observations, used_mask),
-            )
+                assert self._init_subspaces is not None, "the subspaces have not been initialized"
+                assert self._tags is not None
```

Comment:
I think there is an existing issue here that has been highlighted by this.

Contrary to what we discussed previously, `filter_datasets` does need to handle the case when `state` is `None`, as it can be called first by bayesian/ask_tell optimizers; which is what you are doing here. However, in this case we also want to support the case where `self._init_subspaces` is None. The derived classes (e.g. `BatchTrustRegionBox`) handle this in `acquire` by creating some default subspaces; but they don't do this if `filter_datasets` is called first.

So I think we need a way (in the derived classes) to also create default subspaces in `filter_datasets`.

Reply:
One issue is that acquire is passed the search space while filter_datasets isn't.

Can you describe the workflow that is currently broken? AFAICT filter_datasets is *always* called first, right?